hack/generator/pkg/codegen/pipeline_create_arm_types.go

```diff
@@ -13,275 +13,118 @@ import (
-		},
-		// Other defs handler
-		func(def astmodel.TypeDefinition) (astmodel.TypeDefinition, error) {
-			armDef, err := createArmTypeDefinition(
-				definitions,
-				false, // not Spec type
-				def,
-				idFactory)
-			if err != nil {
-				return astmodel.TypeDefinition{}, err
-			}
-
-			kubeNameToArmDefs[def.Name()] = armDef
-
-			return armDef, nil
-		})
-	if err != nil {
-		return nil, nil, err
-	}
-
-	return armDefs, kubeNameToArmDefs, nil
-}
-
-func modifyKubeTypes(
-	definitions astmodel.Types,
-	kubeNameToArmDefs astmodel.Types,
-	idFactory astmodel.IdentifierFactory) (astmodel.Types, error) {
-
-	return iterDefs(
-		definitions,
-		// Resource handler
-		func(name astmodel.TypeName, resourceType *astmodel.ResourceType) ([]astmodel.TypeName, []astmodel.TypeDefinition, error) {
-			specDef, err := modifyKubeResourceSpecDefinition(definitions, idFactory, resourceType)
-			if err != nil {
-				return nil, nil, errors.Wrapf(err, "unable to modify kube resource spec definition for resource %s", name)
-			}
-
-			armSpecDef, ok := kubeNameToArmDefs[specDef.Name()]
-			if !ok {
-				return nil, nil, errors.Errorf("couldn't find arm def matching kube def %q", specDef.Name())
-			}
-
-			modifiedSpec, err := addArmConversionInterface(specDef, armSpecDef, idFactory, armconversion.SpecType)
-			if err != nil {
-				return nil, nil, err
-			}
-
-			names := []astmodel.TypeName{modifiedSpec.Name()}
-			defs := []astmodel.TypeDefinition{modifiedSpec}
-
-			statusType := astmodel.IgnoringErrors(resourceType.StatusType())
-			if statusType != nil {
-				statusName, ok := statusType.(astmodel.TypeName)
-				if !ok {
-					return nil, nil, errors.Errorf("expected Status type to be a name, instead was %T", statusType)
-				}
-
-				statusDef, ok := definitions[statusName]
-				if !ok {
-					return nil, nil, errors.Errorf("could not find Status type %v", statusName)
-				}
-
-				armStatusDef, ok := kubeNameToArmDefs[statusName]
-				if !ok {
-					return nil, nil, errors.Errorf("could not find ARM Status type %v", statusName)
-				}
-
-				newStatus, err := addArmConversionInterface(statusDef, armStatusDef, idFactory, armconversion.StatusType)
-				if err != nil {
-					return nil, nil, err
-				}
-
-				names = append(names, newStatus.Name())
-				defs = append(defs, newStatus)
-			}
-
-			return names, defs, nil
-		},
-		// Other defs handler
-		func(def astmodel.TypeDefinition) (astmodel.TypeDefinition, error) {
-			armDef, ok := kubeNameToArmDefs[def.Name()]
-			if !ok {
-				return astmodel.TypeDefinition{}, errors.Errorf("couldn't find arm def matching kube def %q", def.Name())
-			}
-
-			modifiedDef, err := addArmConversionInterface(def, armDef, idFactory, armconversion.OrdinaryType)
-			if err != nil {
-				return astmodel.TypeDefinition{}, errors.Wrapf(err, "failed to add ARM conversion interface to %q", def.Name())
-			}
-
-			return modifiedDef, nil
-		})
+type armTypeCreator struct {
```

Comment:
This works out much nicer I think!

Reply:
@Porges actually suggested it a while back - agree it's significantly cleaner.